### PR TITLE
New AWS module rds_option_group.py

### PR DESCRIPTION
**Pull request by naslanidis**
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

rds_option_group
##### ANSIBLE VERSION

```
ansible 2.3.0 (devel c5d3aac822) last updated 2017/01/15 12:44:25 (GMT +1000)
```
##### SUMMARY

This module manages the creation, modification, deletion of RDS Option groups and the associated options applied to those groups. 

Note: Some of the options list output has been removed for the sake of brevity. 

```
TASK [print new option group] **************************************************
ok: [localhost] => {
    "msg": {
        "changed": true, 
        "option_group": {
            "OptionGroupsList": [
                {
                    "AllowsVpcAndNonVpcInstanceMemberships": false, 
                    "EngineName": "mysql", 
                    "MajorEngineVersion": "5.6", 
                    "OptionGroupDescription": "test mysql option group", 
                    "OptionGroupName": "test-mysql-option-group", 
                    "Options": [
                        {
                            "DBSecurityGroupMemberships": [], 
                            "OptionDescription": "Innodb Memcached for MySQL", 
                            "OptionName": "MEMCACHED", 
                            "OptionSettings": [
                                {
                                    "AllowedValues": "1-2", 
                                    "ApplyType": "STATIC", 
                                    "DataType": "FLOAT", 
                                    "DefaultValue": "1.25", 
                                    "Description": "Chunk size growth factor that controls the size of each successive chunk with each chunk growing times this amount larger than the previous chunk.", 
                                    "IsCollection": false, 
                                    "IsModifiable": true, 
                                    "Name": "CHUNK_SIZE_GROWTH_FACTOR", 
                                    "Value": "1.35"
                                }, 
                                {
                                    "AllowedValues": "10-1024", 
                                    "ApplyType": "STATIC", 
                                    "DataType": "INTEGER", 
                                    "DefaultValue": "1024", 
                                    "Description": "Maximum number of concurrent connections. Setting this value to anything less than 10 prevents MySQL from starting.", 
                                    "IsCollection": false, 
                                    "IsModifiable": true, 
                                    "Name": "MAX_SIMULTANEOUS_CONNECTIONS", 
                                    "Value": "20"
                                }
                            ], 
                            "Permanent": false, 
                            "Persistent": false, 
                            "Port": 11211, 
                            "VpcSecurityGroupMemberships": [
                                {
                                    "Status": "active", 
                                    "VpcSecurityGroupId": "sg-d188c123"
                                }
                            ]
                        }
                    ], 
                    "VpcId": "vpc-aad24123"
                }
            ], 
            "ResponseMetadata": {
                "HTTPStatusCode": 200, 
                "RequestId": "cb9bf613-7027-11e6-8515-d3586c4629c2"
            }
        }
    }
}
```

### CODE REVIEW

**Comment by naslanidis** [comment]
Migrated from https://github.com/ansible/ansible-modules-extras/pull/2826

**Comment by willthames** [comment]
Once tests pass this is good to go.

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 1]
```diff
@@ -381,4 +381,4 @@ def main():
 from ansible.module_utils.ec2 import *
 
 if __name__ == '__main__':
-    main()
+    main()
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py]: There should be a newline at the end of the file

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        options=dict(type='list', required=False),
+        apply_immediately=dict(type='bool', required=False),
+        state=dict(type='str', default='present', choices=['present', 'absent']),
+        ),
+    )
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='json and boto3 is required.')
+    state = module.params.get('state').lower()
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py]: no need for `lower()` here

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+
+    if existing_option_group:
+        results = existing_option_group
+        if module.params.get('options'):
+            # Check if existing options require updating
+            update_required = match_option_group_options(client, module)
+            # Check if there are options to be added or removed
+            to_be_added, to_be_removed = compare_option_group(client, module)
+            if to_be_added or update_required:
+                changed, new_option_group_options =  create_option_group_options(client, module)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py:294]: Shouldn't this take `to_be_added` ?

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+    to_be_removed = None
+    for current_option in current_option_group['OptionGroupsList']:
+        # Catch new settings you've provided that aren't in the current settings
+        old_settings = []
+        new_settings = []
+        for setting_name in new_options:
+            new_settings.append(setting_name['OptionName'])
+        for option in current_option['Options']:
+            old_settings.append(option['OptionName'])
+        if cmp(sorted(old_settings),sorted(new_settings)) != 0:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py]: This whole section would be much simpler if you started with `set`s and continued with them

```
new_settings = set([item['OptionName'] for item in new_options])
old_settings = set([item['OptionName'] for item in current_option['Options'])
if new_settings != old_settings:
    to_be_added = list(new_settings - old_settings)
    to_be_removed = list(old_settings - new_settings)
```

**Comment by willthames** [comment]
Other than the final newline, all my review comments are more suggestions than requirements.

**Comment by naslanidis** [comment]
@willthames 

Your suggestions were good and I've added all but one. I just retested and the to_be_added doesn't get passed to the create_option_group_function as if there are any new ones at all it just gets them from the original params. The to_be_removed gets passed to the remove function as we only want to remove specific options.

File: lib/ansible/modules/cloud/amazon/rds_option_group.py (hunk 3 above)
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py]: You inadvertently added a third blank line. Probably worth running flake8 and removing what issues you can

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        new_settings = set([item['OptionName'] for item in new_options])
+        old_settings = set([item['OptionName'] for item in current_option['Options']])
+        if new_settings != old_settings:
+            to_be_added = list(new_settings - old_settings)
+            to_be_removed = list(old_settings - new_settings)
+
+        return to_be_added, to_be_removed
+
+
+
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py:278]: You inadvertently added a third blank line. Probably worth running flake8 and removing what issues you can

Comment (willthames) [reply lib/ansible/modules/cloud/amazon/rds_option_group.py:278]: If that's too much of a headache, even just removing the accidental line would be good.

**Comment by willthames** [comment]
shipit

Extra blank line is not a big deal in the grand scheme of things

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+
+DOCUMENTATION = '''
+module: rds_option_group
+short_description: Manages RDS Option Groups
+description:
+  -Manages the creation, modification, deletion of RDS option groups.
+version_added: "2.3"
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py:23]: Should be 2.4 now.

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+    invocations = {
+        "present": setup_rds_option_group,
+        "absent": remove_rds_option_group
+    }
+
+    (changed, results) = invocations[state](client, module)
+    module.exit_json(changed=changed, option_group=results)
+
+# Import module snippets
+from ansible.module_utils.basic import *
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py:380]: Import just the specific things used and move these to the top, please!

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+    except botocore.exceptions.NoCredentialsError:
+        e = get_exception()
+        module.fail_json(msg="Can't authorize connection - "+str(e))
+
+    invocations = {
+        "present": setup_rds_option_group,
+        "absent": remove_rds_option_group
+    }
+
+    (changed, results) = invocations[state](client, module)
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py:376]: Parentheses aren't needed here.

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+            requires_update = True
+        else:
+            for option in current_option['Options']:
+                for setting_name in new_options:
+                    if setting_name['OptionName'] == option['OptionName']:
+                        for name in iter(setting_name):
+                            # Security groups need to be handled separately due to different keys on request and what is
+                            # returned by the API
+                            if name in iter(option) and name != 'OptionSettings' and name != 'VpcSecurityGroupMemberships':
+                                if cmp(setting_name[name], option[name]) != 0:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py:242]: No cmp with python3.

File: lib/ansible/modules/cloud/amazon/rds_option_group.py [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+                                for groups in option[name]:
+                                    if groups['VpcSecurityGroupId'] not in setting_name[name]:
+                                        requires_update = True
+
+                    for new_option in new_options:
+                        if option['OptionName'] == new_option['OptionName']:
+                            for current_option_setting in option['OptionSettings']:
+                                for new_option_setting in new_option['OptionSettings']:
+                                    if new_option_setting['Name'] == current_option_setting['Name']:
+                                        if cmp(new_option_setting['Value'], current_option_setting['Value']) != 0:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/rds_option_group.py:254]: Here too.

**Review by s-hertel — CHANGES_REQUESTED** [review]
After these things get addressed this gets a shipit from me too. Looks good!

**Comment by willthames** [comment]
In the absence of further updates from @naslanidis I've implemented the fixes for this PR in #25290.

**Comment by jctanner** [comment]
duplicate_of https://github.com/ansible/ansible/pull/25290

**Pull request closed without merging**